src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,230 @@
+					DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsMethodReturnType :
+					DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsThisParameter,
+				(SymbolKind.Method, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsField,
+				(SymbolKind.Method, SymbolKind.Method) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsMethodReturnType :
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsThisParameter,
+				(SymbolKind.Method, SymbolKind.Parameter) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsParameter :
+					DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsParameter,
+				(SymbolKind.NamedType, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsField,
+				(SymbolKind.TypeParameter, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsField,
+				(SymbolKind.TypeParameter, SymbolKind.Method) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsMethodReturnType,
+				(SymbolKind.TypeParameter, SymbolKind.Parameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsParameter,
+				(SymbolKind.TypeParameter, SymbolKind.TypeParameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsGenericParameter,
+				_ => throw new NotImplementedException ()
+			};
+
+		static string[] GetDiagnosticArguments (ISymbol source, ISymbol target, string missingAnnotations)
+		{
+			var args = new List<string> ();
+			args.AddRange (GetDiagnosticArguments (target));
+			args.AddRange (GetDiagnosticArguments (source));
+			args.Add (missingAnnotations);
+			return args.ToArray ();
+		}
+
+		static IEnumerable<string> GetDiagnosticArguments (ISymbol symbol)
+		{
+			var args = new List<string> ();
+			args.AddRange (symbol.Kind switch {
+				SymbolKind.Parameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				SymbolKind.NamedType => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Field => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Method => new string[] { symbol.GetDisplayName () },
+				SymbolKind.TypeParameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				_ => throw new NotImplementedException ($"Unsupported source or target symbol {symbol}.")
+			});
+
+			return args;
+		}
+
+		static ISymbol? TryGetSymbolFromOperation (IOperation? operation) =>
+			operation switch {
+				IArgumentOperation argument => TryGetSymbolFromOperation (argument.Value),
+				IAssignmentOperation assignment => TryGetSymbolFromOperation (assignment.Value),
+				IConversionOperation conversion => conversion.Type,
+				IInstanceReferenceOperation instanceReference => instanceReference.Type,
```

Comment:
Maybe leave a comment here that explains we are using the type of the instance to represent `this` values which gets its annotations from the calling method.